internal/client/client_batch.go

```diff
@@ -366,37 +380,49 @@ func (a *batchConn) getClientAndSend() {
 		cli    *batchCommandsClient
 		target string
 	)
+	reason := ""
 	for i := 0; i < len(a.batchCommandsClients); i++ {
 		a.index = (a.index + 1) % uint32(len(a.batchCommandsClients))
 		target = a.batchCommandsClients[a.index].target
 		// The lock protects the batchCommandsClient from been closed while it's in use.
-		if a.batchCommandsClients[a.index].tryLockForSend() {
-			cli = a.batchCommandsClients[a.index]
-			break
+		if c := a.batchCommandsClients[a.index]; c.tryLockForSend() {
+			if c.sent.Load() <= c.maxConcurrencyRequestLimit.Load() {
+				cli = c
+				break
+			} else {
+				reason = SendFailedReasonNoAvailableLimit
+				c.unlockForSend()
```

Comment:
```suggestion
				reason = SendFailedReasonNoAvailableLimit
				c.unlockForSend()
				c = nil
```

Reply:
Maybe it's better to first check the request limit and then lock a client

Reply:
good idea